core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java

```diff
@@ -81,6 +81,7 @@
     private double tableScanNodePartitioningMinBucketToTaskRatio = 0.5;
     private boolean mergeProjectWithValues = true;
     private boolean forceSingleNodeOutput = true;
+    private boolean useExactPartitioning;
```

Comment:
Did we discuss what should be default value for this one?
I would keep it `false` for now but we should consider changing it to `true`. @linzebing can you spin up performance tests to compare results for both options?

Reply:
I feel it's risky to default to true. Ideally Trino should make smarter decisions at runtime on whether to use exact partitioning. This PR intends only to provide a config to force exact partitioning.